File: server/deploy/currentVersion/pookman/scripts/game.js
/*

CHANGES:
-----------------------------------
30-3-16 : 	Speed for pacman and ghosts in every mode setActive
			CanWalk bugFix : Not more wallthrough!

*/


var	ACTION_DEBUG_TEST_INVENTORY	= false,
	ACTION_DEBUG_NEXT_SONG 		= false,
	ACTION_DEBUG_ABORT 			= false,
	ACTION_ACTIVATE 			= false,
	ACTION_NEXT_POWERUP 		= false,
	ACTION_PREV_POWERUP 		= false,
	ACTION_MOVE_LEFT 			= false,
	ACTION_MOVE_RIGHT 			= false,
	ACTION_MOVE_UP 				= false,
	ACTION_MOVE_DOWN 			= false,
	ACTION_CANCEL 				= false,
	ACTION_PAUSE 				= false;

var debug = [];
var	DEBUG_PACMAN 	= 0x00,
	DEBUG_GHOSTS 	= 0x01,
	DEBUG_SLOWMO 	= 0x02,
	DEBUG_GAME		= 0x03;
	
debug[DEBUG_PACMAN] = false;
debug[DEBUG_GHOSTS] = false;
debug[DEBUG_GAME] 	= false;
debug[DEBUG_SLOWMO] = false;


var DOOR_X = [13, 14];
var DOOR_Y = 12;


function main(){
	info('JSEngine', 'Initializing');
	
	game = new Game(224, 288, true, 1);
}

Game.prototype.addScore = function(score){	
	this.score += score;
	$('#score').html((("00000000"+(this.score.toString())).substr(-8)));
}

Game.prototype.continueLoad = function(){	
	try {
		setTimeout(update, 15);
		cout("^6------------------- Starting new game --------------------");
	} catch (e){
		complaint('GAME', 'EXCEPTION RAISED', e);
	}
}

Game.prototype.resetInput = function(){ 
	ACTION_DEBUG_ABORT 		= false;
	ACTION_DEBUG_NEXT_SONG 	= false;
	ACTION_CANCEL 			= false;
	ACTION_PAUSE 			= false;        
}

Game.prototype.handleInput = function(){
	//if(this.demo.playback)return;
	this.engine.input.update();
	this.resetInput();
	var input = this.engine.input;
	
	// Handle movement
	if(input.keyPress[input.KEY_A]) this.pacman.go(DIR_LEFT);	else 
	if(input.keyPress[input.KEY_D]) this.pacman.go(DIR_RIGHT);	else 
	if(input.keyPress[input.KEY_W]) this.pacman.go(DIR_UP);		else 
	if(input.keyPress[input.KEY_S]) this.pacman.go(DIR_DOWN);	
			
	if(input.keyDown[input.KEY_X]) debug[DEBUG_SLOWMO] ^= 1;
	
	if(input.keyDown[input.KEY_N]) this.pacman.spawn();	
	
	if(input.keyDown[input.KEY_P]		) debug[DEBUG_PACMAN] ^= 1;		else 	
	if(input.keyDown[input.KEY_O]		) debug[DEBUG_GAME] ^= 1;		else 	
	if(input.keyDown[input.KEY_I]		) debug[DEBUG_GHOSTS] ^= 1;		else 	
	
	
	if(input.keyDown[input.KEY_1]		) this.setMode(MODE_GAME_ARCADE);		else 
	if(input.keyDown[input.KEY_2]		) this.setMode(MODE_GAME_TIMEATTACK); 	else 
	if(input.keyDown[input.KEY_3]		) this.setMode(MODE_GAME_PUZZLE); 		else 
	if(input.keyDown[input.KEY_4]		) this.setMode(MODE_GAME_VERSUS); 		else
	if(input.keyDown[input.KEY_G]		) this.setMode(MODE_CURTAIN_GAMEOVER);	else 
	if(input.keyDown[input.KEY_M]		) this.setMode(MODE_MENU);				else 
	if(input.keyDown[input.KEY_END]		) ACTION_DEBUG_ABORT 		= true;	
	//if(input.keyDown[input.KEY_N]    	) ACTION_DEBUG_NEXT_SONG  	= true;
	if(input.keyDown[input.KEY_F1]    	) this.nextMap();

	if(input.keyDown[input.KEY_RIGHT]){	ACTION_BALLTIME 	= true; 
										ACTION_FORCE_TILT 	= true; 
										ACTION_TILT_RIGHT 	= true; } else
	if(input.keyDown[input.KEY_LEFT] ){	ACTION_BALLTIME 	= true; 
										ACTION_FORCE_TILT 	= true; 
										ACTION_TILT_LEFT  	= true; }	
	if(input.keyDown[input.KEY_DOWN] ){	ACTION_BALLTIME 	= true; 
										ACTION_FORCE_TILT 	= true; 
										ACTION_TILT_DOWN 	= true; } else
	if(input.keyDown[input.KEY_UP]   ){	ACTION_BALLTIME 	= true; 
										ACTION_FORCE_TILT 	= true; 
										ACTION_TILT_UP   	= true; }
	
	if(input.keyDown[input.KEY_SPACE] ) ACTION_ACTIVATE = true;
	if(input.keyDown[input.KEY_ESCAPE]) ACTION_CANCEL   = true;
}

Game.prototype.loadGraphics = function(){
	new Menu().loadGraphics();
	new ParticleSystem().loadGraphics();
	
	for(var i=0; i<FONT_FILES.length; i++){
		new Display(0,0,0,0,i).loadGraphics();
	}
}

var	MODE_GAME_ARCADE 		= 10,
	MODE_GAME_TIMEATTACK 	= 11,
	MODE_GAME_VERSUS 		= 12,
	MODE_GAME_PUZZLE 		= 13,
	MODE_CURTAIN_ROUNDCLEAR = 20,
	MODE_CURTAIN_BONUSCOINS = 21,
	MODE_CURTAIN_ENTERLEVEL = 22,
	MODE_CURTAIN_GAMEOVER 	= 23,
	MODE_MENU 				= 0,
	MODE_EXIT 				= 100;

function Game(width, height, fullScreen, scale){
	this.engine 	= new Engine("Game", width, height, fullScreen, scale);

	this.graphics 	= this.engine.graphics;	
	this.particles 	= this.engine.particles;			
    this.accel 		= this.engine.accel;
	this.maxAccel 	= this.engine.maxAccel;
	this.input 		= this.engine.input;
	this.graphics	= this.engine.graphics;

	this.player 	= "anonymous";	
    this.gameOver 	= false;
	this.cheats 	= false;
    this.map 		= false;	
	this.menu 		= false;	
    this.lives 		= 3;
    this.level 		= 0;
    this.score 		= 0;
	this.gameMode 	= 0;
	this.pills		= 0;
	this.pillsLeft 	= 0;

	this.mapWidth	= 0;
	this.mapHeight	= 0;
	
	this.music		= new MusicPlayer(this);
	this.menu 		= new Menu(this);
	this.demo 		= new Demo(this);
	this.bonus		= new Bonus(this);
	this.pacman		= new Pacman(this);
	this.blinky		= new Ghost(GHOST_BLINKY, this);
	this.inky		= new Ghost(GHOST_INKY, this);
	this.pinky		= new Ghost(GHOST_PINKY, this);
	this.clyde		= new Ghost(GHOST_CLYDE, this);
	this.map 		= new Map(MAP_LIST[0],this);
	
	this.freeze		= 0;
	
    this.mode 		= MODE_GAME_ARCADE;
    this.status 	= 0;
    
	resources.wait(this.continueLoad, this);
}
     
Game.prototype.resetSystems = function(){
	this.engine.particles.reset();
	this.blinky.reset();
	this.pinky.reset();
	this.inky.reset();
	this.clyde.reset();
	this.pacman.ready();	
}

Game.prototype.New = function() {
	cout("^6------------------- Starting new game --------------------");

	this.gameOver 	= false;
	this.cheats 	= false;
    this.map 		= false;	
	this.menu 		= false;	
    this.lives 		= 0;
    this.level 		= 0;
    this.score 		= 0;
	this.gameMode 	= 0;
	this.engine.timeScale = 1.0;
	this.resetSystems();

	this.countPills();
	this.mapWidth	= this.map.map.tileWidth 	* this.map.map.width;
	this.mapHeight	= this.map.map.tileHeight	* this.map.map.height;
	
	
	this.bonus.reset();
	
	this.snapshot();
	
	cout("New game started");		
}   

Game.prototype.closeHouse = function(){
	this.map.set(DOOR_X[0], DOOR_Y, 0x20);
	this.map.set(DOOR_X[1], DOOR_Y, 0x20);
	this.map.redraw = true;
}	

Game.prototype.openHouse = function(){
	this.map.set(DOOR_X[0], DOOR_Y, 0x1F);
	this.map.set(DOOR_X[1], DOOR_Y, 0x1F);
	this.map.redraw = true;
}	
	
Game.prototype.nextMap = function(){
	this.resetSystems();
	this.level = this.level + 1;
	//this.musicplayer.nextTrack();
	this.map.loadMap(MAP_LIST[this.level % (MAP_LIST.length-1)]);
	this.pacman.spawn();
	this.blinky.reset();
	this.inky.reset();
	this.pinky.reset();
	this.clyde.reset();
	this.closeHouse();
	this.bonus.reset();	
}

Game.prototype.setMode = function(mode){
	this.mode = mode;
	this.status = 0;
	this.timeout = 0;
}

Game.prototype.oneUp = function(lives){
	this.lives += lives;        
}

Game.prototype.oneDown = function(lives){
	this.lives -= lives;
	if( this.lives <= 0 ) {
		this.lives = 0;
		this.gameOver = true;		
	}
}
 
Game.prototype.isGhost = function(i) {
	if(this.pacman.status == PACMAN_DIEING) return false;
	else if(this.pacman.status == PACMAN_DEAD) return false;
	
	var w = this.map.map.width;
	var h = this.map.map.height;
	var pos = [];
	
	pos[GHOST_BLINKY]	= (this.blinky.ty	* w) + this.blinky.tx;
	pos[GHOST_PINKY]	= (this.pinky.ty 	* w) + this.pinky.tx;
	pos[GHOST_INKY] 	= (this.inky.ty 	* w) + this.inky.tx;
	pos[GHOST_CLYDE]	= (this.clyde.ty 	* w) + this.clyde.tx;
	
	return	((this.blinky.mode	< STATUS_RETURNING)&&(pos[GHOST_BLINKY] == i))||
			((this.pinky.mode	< STATUS_RETURNING)&&(pos[GHOST_PINKY ] == i))||
			((this.inky.mode 	< STATUS_RETURNING)&&(pos[GHOST_INKY  ] == i))||
			((this.clyde.mode	< STATUS_RETURNING)&&(pos[GHOST_CLYDE ] == i));
}
	
Game.prototype.isPill = function(i) {	
	return this.map.map.layers[0].data[i] == 33;
}

Game.prototype.isEnergizer = function(i) {	
	return this.map.map.layers[0].data[i] == 34;
}



Game.prototype.countPills = function() {	
	this.pillsLeft = 0;
	for(var i=0; i<this.map.map.layers[0].data.length; i++){
		if(this.map.map.layers[0].data[i] == 33)this.pillsLeft++;
	}
	if(this.pillsLeft==0) {
		this.pillsLeft = -1;
		this.nextMap();
	}
}

Game.prototype.update = function(delta) {	
	this.handleInput();

	
	
	// Freeze game while kill sound is playing
	if(this.freeze <= 0){
		this.bonus.update();
		// Delay pacman a frame when it eats 
		if(!this.pacman.eating){	
			this.pacman.update(delta);
		} else {
			this.countPills();
			this.pacman.eating = false;
		}
		
		this.blinky.update(delta);
		this.inky.update(delta);
		this.pinky.update(delta);
		this.clyde.update(delta);
		if(this.pinky.mustLeave)this.openHouse();
		else this.closeHouse();
		
		if(debug[DEBUG_SLOWMO]) this.freeze = 10;
		
	} else this.freeze--;
	
	// var	t = parseInt(Math.random() * ParticleGfx.length),
	// x = parseInt(Math.random() * this.engine.width),
	// y = parseInt(Math.random() * this.engine.height);	
	//this.engine.particles.update(delta);
}

Game.prototype.render = function() {

	this.engine.graphics.setContext(0);
	
	// Draw lives
	if(this.map.redraw){
	for(var i=0; i<this.lives; i++){
		PacmanGfx[DIR_RIGHT].draw(i<<4, 272);
	}
	
	// Draw current level (bonus)
	for(var i=0, li = (this.level > 2) ? 3 : (this.level+1); i<li; i++){
		BonusGfx.setCurrentFrame(this.level-((li-1)-i));
		BonusGfx.draw(208 - (i<<4), 272);
	}
	}
	
	this.map.render(this.engine.graphics);
	this.bonus.render();
	//this.game.engine.graphics.setContext(1);
	//this.engine.particles.render(this.engine.graphics);
	PacmanGfx[DIR_RIGHT].setCurrentFrame(9);
	
	this.engine.graphics.setContext(2);
	this.engine.graphics.clear();
	this.pacman.render(this.engine.graphics);
	
	this.engine.graphics.setContext(3);
	this.engine.graphics.clear();
	
	if(this.pacman.status != PACMAN_DIEING) {
		this.blinky.render(this.engine.graphics);	
		this.inky.render(this.engine.graphics);
		this.pinky.render(this.engine.graphics);
		this.clyde.render(this.engine.graphics);
	}
}

File: server/deploy/currentVersion/pookman/scripts/game.py
#################################################################################################################
#                                                                                                               #
# 88888888ba     ,ad8888ba,      ,ad8888ba,    88      a8P   88b           d88         db         888b      88  #
# 88      "8b   d8"'    `"8b    d8"'    `"8b   88    ,88'    888b         d888        d88b        8888b     88  #
# 88      ,8P  d8'        `8b  d8'        `8b  88  ,88"      88`8b       d8'88       d8'`8b       88 `8b    88  #
# 88aaaaaa8P'  88          88  88          88  88,d88'       88 `8b     d8' 88      d8'  `8b      88  `8b   88  #
# 88""""""'    88          88  88          88  8888"88,      88  `8b   d8'  88     d8YaaaaY8b     88   `8b  88  #
# 88           Y8,        ,8P  Y8,        ,8P  88P   Y8b     88   `8b d8'   88    d8""""""""8b    88    `8b 88  #
# 88            Y8a.    .a8P    Y8a.    .a8P   88     "88,   88    `888'    88   d8'        `8b   88     `8888  #
# 88             `"Y8888Y"'      `"Y8888Y"'    88       Y8b  88     `8'     88  d8'          `8b  88      `888  #
#                                                                                                               #
#################################################################################################################
from data.scripts.map           import map
from data.scripts.pacmanstatus  import PacmanStatus
from data.scripts.ghoststatus   import GhostStatus
from data.scripts.ghosttype     import GhostType
from data.scripts.gamemode      import GameMode
from data.scripts.pacman        import Pacman
from data.scripts.ghosts        import Ghost
from data.scripts.bonus         import Bonus
from scripts.main               import menu
from tiledmap                   import TiledMap
from random                     import random
from direction                  import *
import blackbox
import vpu
import joypad

class Debug:
    PACMAN  = False
    GHOSTS  = False
    GAME    = True
    SLOWMO  = False

class Action:
    DEBUG_TEST_INVENTORY= False
    DEBUG_NEXT_SONG     = False
    DEBUG_ABORT         = False
    ACTIVATE            = False
    NEXT_POWERUP        = False
    PREV_POWERUP        = False
    MOVE_LEFT           = False
    MOVE_RIGHT          = False
    MOVE_UP             = False
    MOVE_DOWN           = False
    CANCEL              = False
    PAUSE               = False

class Door:
    x       = [13, 14]
    y       = 12
    open    = False

class Game:
    debug   = Debug
    action  = Action
    running = True
    width = 0
    height = 0
    dims = {}
    scale = 2.0
    score = 0
    time = 0
    
    level   = 0
    lives   = 0
    mode    = 0
    status  = 0
    timeout = 0
    gameOver=False
    cheats  =False
    gameMode=False
    freeze = False
    
    map_width  = 28 * 8 
    map_height = 31 * 8

    blinky = None
    pinky  = None
    inky   = None
    clyde  = None
    pacman = None
    bonus  = None

    buffer = [ None, None ]
    
    map = None

    door = Door

    def resetInput(): 
        Game.action.DEBUG_ABORT = False
        Game.action.NEXT_SONG     = False
        Game.action.CANCEL         = False
        Game.action.PAUSE         = False

    @staticmethod
    def resetSystems():
        Game.blinky.reset()
        Game.pinky.reset()
        Game.inky.reset()
        Game.clyde.reset()
        Game.pacman.ready()    

    @staticmethod
    def closeHouse():
        print("Closing House")
        Game.door.open = False
        Game.map.set(Game.door.x[0], Game.door.y, 0x1F)
        Game.map.set(Game.door.x[1], Game.door.y, 0x1F)
        Game.map.redraw()

    @staticmethod
    def openHouse():
        print("Opening House")
        Game.door.open = True
        Game.map.set(Game.door.x[0], Game.door.y, 0x3F)
        Game.map.set(Game.door.x[1], Game.door.y, 0x3F)
        Game.map.redraw()
    
    @staticmethod
    def nextMap():
        Game.resetSystems()
        Game.level = Game.level + 1
        #reload map!
        Game.pacman.spawn()
        Game.blinky.reset()
        Game.inky.reset()
        Game.pinky.reset()
        Game.clyde.reset()
        Game.closeHouse()
        Game.bonus.reset()
        Game.reload_map()

    @staticmethod
    def setMode(mode):
        Game.mode = mode
        
        
    @staticmethod
    def oneUp(lives):
        Game.lives += lives

    @staticmethod
    def oneDown(lives):
        Game.lives -= lives
        if Game.lives <= 0 :
            Game.lives = 0
            Game.gameOver = True
        
    @staticmethod
    def is_pill(i):
        x = i % 28
        y = int(i / 28)
        return Game.map.get(x,y,0) == 0x20
    
    @staticmethod
    def is_energizer(i):
        x = i % 28
        y = int(i / 28)
        return Game.map.get(x,y,0) == 0x21

    @staticmethod 
    def countPills():
        Game.pillsLeft = Game.map.count(0x20)
        if Game.pillsLeft==0:
            Game.pillsLeft = -1
            Game.nextMap()

    @staticmethod 
    def New():
        print("^6------------------- Starting new game --------------------")
        Game.gameOver             = False
        Game.cheats             = False
        Game.map                 = False    
        Game.menu                 = False    
        Game.lives                 = 0
        Game.level                 = 0
        Game.score                 = 0
        Game.gameMode             = 0
        Game.resetSystems()
        Game.countPills()
        Game.bonus.reset()
        Game.snapshot()
        Game.reload_map()        
        print("New game started")

    @staticmethod
    def is_ghost(i):
        if Game.pacman.status == PacmanStatus.DIEING: return False
        elif Game.pacman.status == PacmanStatus.DEAD: return False
        
        w = Game.map.width
        h = Game.map.height
        pos = {
            GhostType.BLINKY    : (Game.blinky.ty * w) + Game.blinky.tx,
            GhostType.PINKY     : (Game.pinky.ty  * w) + Game.pinky.tx,
            GhostType.INKY      : (Game.inky.ty   * w) + Game.inky.tx,
            GhostType.CLYDE     : (Game.clyde.ty  * w) + Game.clyde.tx,
        }
        return ((Game.blinky.mode    < GhostStatus.RETURNING) and (pos[GhostType.BLINKY] == i)) or ((Game.pinky.mode    < GhostStatus.RETURNING) and (pos[GhostType.PINKY ] == i)) or ((Game.inky.mode     < GhostStatus.RETURNING) and (pos[GhostType.INKY  ] == i)) or ((Game.clyde.mode    < GhostStatus.RETURNING) and (pos[GhostType.CLYDE ] == i))

    @staticmethod
    def reload_map():
        Game.map.fill(0x6f)
        #load map...
        i=0
        for y in range(0, 31):
            for x in range(0, 28):
                Game.map.setvalue(x,y, map[i], 0)
                i+=1

    @staticmethod
    def setup():
        print("GAME: Setting up...")
        Game.running = True
        Game.map = map
        # disable rendering
        vpu.disable(0)
        vpu.disable(1)
        vpu.disable(2)
        # set video scale to 2x
        vpu.setscale(0, 2.0, 2.0)
        vpu.setscale(1, 2.0, 2.0)        
        vpu.setscale(2, 1.0, 1.0) # Overlay is always drawn at 2x zomm
        vpu.setrotation(0, 0.0)
        vpu.setrotation(1, 0.0)        
        vpu.setrotation(2, 0.0)  
        vpu.setfont('tiny')
        # get layer dimensions
        vpu.select(0); Game.dims[0] = vpu.dimensions()
        Game.width  = int(Game.dims[0][0])
        Game.height = int(Game.dims[0][1])
        scale_y = Game.height / (31*8)
        vpu.setscale(0, 2, scale_y/2)
        vpu.select(1); Game.dims[1] = vpu.dimensions()
        vpu.select(2); Game.dims[2] = vpu.dimensions()
        
        #allocate custom video buffers
        Game.buffer[0] = vpu.createsurf(Game.map_width, Game.map_height)
        Game.buffer[1] = vpu.createsurf(Game.map_width, Game.map_height)
        
        # prepare initial layer state
        vpu.select(0); vpu.fill(0,0,0,255)
        vpu.select(1); vpu.fill(0,0,0,255)
        vpu.select(2); vpu.fill(0,0,0,0)
        vpu.select(Game.buffer[0]); vpu.fill(0,0,0,0)
        vpu.select(Game.buffer[1]); vpu.fill(0,0,0,0)
        
        print("GAME: Initializing classes...")
        Game.map = TiledMap(Game, 28, 31)
        if not Game.map.load_tileset("walls"):
            print("Cannot load tileset!")
                

        Game.reload_map()
        
        Game.pacman = Pacman
        Game.bonus  = Bonus
        Game.ghost  = Ghost
        #Game.demo     = Demo
        #Game.music = MusicPlayer
        
        Game.pacman.initialize(Game)
        Game.ghost.initialize(Game)
        Game.bonus.initialize(Game)
        #Game.demo.initialize(Game)
        #Game.music.initialize(Game)

        Game.blinky     = Ghost(GhostType.BLINKY)
        Game.inky       = Ghost(GhostType.INKY  )
        Game.pinky      = Ghost(GhostType.PINKY )
        Game.clyde      = Ghost(GhostType.CLYDE )

        Game.gameOver   = False
        Game.cheats     = False
        Game.lives      = 3
        Game.level      = 0
        Game.score      = 0
        Game.gameMode   = 0
        Game.pills      = 0
        Game.pillsLeft  = 0

        Game.freeze     = 0
        
        Game.mode       = GameMode.ARCADE
        
        
        # enable rendering back
        vpu.enable(0)
        #vpu.enable(1)
        vpu.enable(2)
        
    @staticmethod
    def destroy():
        Ghost.destroy()
        Pacman.destroy()
        Bonus.destroy()

        vpu.select(0); vpu.fill(0,0,0,255)
        vpu.select(1); vpu.fill(0,0,0,255)
        vpu.select(2); vpu.fill(0,0,0,0)        

    @staticmethod
    def loop():
        delta = 1.0
        while Game.running:
            Game.draw()
            Game.update(delta)            
            
    @staticmethod
    def update_input(delta):
        #if Game.demo.playback: return
        
        # Handle movement
        if joypad.left()    : Game.pacman.go(Direction.LEFT)
        elif joypad.right() : Game.pacman.go(Direction.RIGHT) 
        elif joypad.up()    : Game.pacman.go(Direction.UP)      
        elif joypad.down()  : Game.pacman.go(Direction.DOWN)    
                
        Game.debug.SLOWMO = True if joypad.select() else False
        
        if joypad.start()   : Game.pacman.spawn()    
        
        #if(input.keyDown[input.KEY_P]        ) debug[DEBUG_PACMAN] ^= 1; else     
        #if(input.keyDown[input.KEY_O]        ) debug[DEBUG_GAME] ^= 1;   else     
        #if(input.keyDown[input.KEY_I]        ) debug[DEBUG_GHOSTS] ^= 1; else     
        
    @staticmethod
    def update(delta):
        Game.update_input(delta)
        # Freeze game while kill sound is playing
        if Game.freeze <= 0:
            Game.bonus.update(delta)
            # Delay pacman a frame when it eats 
            if not Game.pacman.eating:
                Game.pacman.update(delta)
            else:
                Game.countPills()
                Game.pacman.eating = False
            
            Game.blinky.update(delta)
            Game.inky.update(delta)
            Game.pinky.update(delta)
            Game.clyde.update(delta)
            
            if Game.pinky.mustLeave:
                if not Game.door.open: 
                    Game.openHouse()

            Game.map.redraw()
            if Game.debug.SLOWMO:
                Game.freeze = 10
        else:
            Game.freeze-=1
        
        #Game.map.update(delta)

        Game.time+=1
        if Game.time % 16==0:
            Game.map.redraw()

        # required stuff
        if blackbox.ctrlc():
            print("Control+C pressed.")
            Game.destroy()
            Game.running = False
            
        if joypad.menu():
            menu()
        
    @staticmethod
    def draw():
        # clear sprite buffer
        vpu.select(Game.buffer[1])
        vpu.fill(0,0,0,0)
        
        if Game.map.need_redraw:
            #clear bg buffer
            vpu.select(Game.buffer[0])
            vpu.fill(0,0,0,0)
            
            # Draw lives
            for i in range(0, Game.lives):
                Game.pacman.gfx[ Direction.RIGHT ].draw(i<<4, 272)
            
            # Draw current level (bonus)
            li = 3 if (Game.level > 2) else (Game.level+1)
            for i in range(0, li):                
                Bonus.gfx[Game.level-((li-1)-i)].draw(208 - (i<<4), 272)

            
            # draw map here
            # --------------------------
            Game.map.x = ( Game.width   >> 1 ) - (( Game.map.width  * Game.map.tile_width  ) >> 1 )
            Game.map.y = ( Game.height  >> 1 ) - (( Game.map.height * Game.map.tile_height ) >> 1 )
            Game.map.draw()
            # --------------------------
            if Bonus.isAlive:
                i = Bonus.game.level % (32*8)
                Bonus.gfx[i].draw(Bonus.x, Bonus.y+24)
        # draw sprites 
        vpu.select(Game.buffer[1])
        vpu.fill(0,0,0,0)
        Pacman.gfx[Direction.RIGHT].setframe(9)
        Game.pacman.draw()
        
        if Game.pacman.status != PacmanStatus.DIEING: 
            Game.blinky.draw()    
            Game.inky.draw()
            Game.pinky.draw()
            Game.clyde.draw()

        vpu.select(0)
        left = (Game.width  >> 1) - (Game.map_width>>1)
        top  = (Game.height >> 1) - (Game.map_height>>1)
        vpu.drawsurf(Game.buffer[0], left, top)
        vpu.drawsurf(Game.buffer[1], left, top)
        vpu.update()

def setup():
    return Game.setup()

def loop():
    return Game.loop()

def destroy():
    return Game.destroy()


